samza-core/src/main/java/org/apache/samza/zk/ZkLeaderElector.java

```diff
@@ -130,6 +130,7 @@ public void tryBecomeLeader() {
         LOG.debug(zLog("Unsubscribing data change for " + currentSubscription));
         zkUtils.unsubscribeDataChanges(keyBuilder.getProcessorsPath() + "/" + currentSubscription,
             previousProcessorChangeListener);
+        previousProcessorChangeListener = new PreviousProcessorChangeListener(zkUtils);
```

Comment:
fyi, I did a pass at the other `IZKDataListener`s for barrier-joins and job-model upgrade as well and how they are instantiated, it lgtm.

Reply:
Yes, this is the only place where the reregistration was done incorrectly.